File: twidere/src/main/java/org/mariotaku/twidere/activity/SettingsActivity.java
/*
 * 				Twidere - Twitter client for Android
 * 
 *  Copyright (C) 2012-2014 Mariotaku Lee <mariotaku.lee@gmail.com>
 * 
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 * 
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 * 
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.mariotaku.twidere.activity;

import android.app.Activity;
import android.app.Dialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.DrawableRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.XmlRes;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.support.v4.widget.SlidingPaneLayout;
import android.support.v7.app.AlertDialog;
import android.support.v7.preference.Preference;
import android.support.v7.preference.PreferenceFragmentCompat;
import android.support.v7.preference.PreferenceFragmentCompat.OnPreferenceDisplayDialogCallback;
import android.support.v7.preference.PreferenceFragmentCompat.OnPreferenceStartFragmentCallback;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AbsListView;
import android.widget.AdapterView;
import android.widget.AdapterView.OnItemClickListener;
import android.widget.BaseAdapter;
import android.widget.ImageView;
import android.widget.ListView;
import android.widget.TextView;

import org.mariotaku.twidere.R;
import org.mariotaku.twidere.activity.support.BaseAppCompatActivity;
import org.mariotaku.twidere.fragment.CustomTabsFragment;
import org.mariotaku.twidere.fragment.ExtensionsListFragment;
import org.mariotaku.twidere.fragment.SettingsDetailsFragment;
import org.mariotaku.twidere.fragment.support.BaseSupportDialogFragment;
import org.mariotaku.twidere.fragment.support.SupportBrowserFragment;
import org.mariotaku.twidere.preference.iface.IDialogPreference;
import org.mariotaku.twidere.util.KeyboardShortcutsHandler;
import org.mariotaku.twidere.util.ThemeUtils;
import org.mariotaku.twidere.util.TwidereActionModeForChildListener;
import org.mariotaku.twidere.view.TintedStatusNativeActionModeAwareLayout;

import java.util.ArrayList;
import java.util.List;

public class SettingsActivity extends BaseAppCompatActivity implements OnItemClickListener,
        OnPreferenceStartFragmentCallback, OnPreferenceDisplayDialogCallback {

    private static final int RESULT_SETTINGS_CHANGED = 10;

    private ListView mEntriesListView;
    private SlidingPaneLayout mSlidingPaneLayout;

    private boolean mShouldNotifyChange;
    private EntriesAdapter mEntriesAdapter;
    private TintedStatusNativeActionModeAwareLayout mMainContent;
    private View mDetailFragmentContainer;
    private TwidereActionModeForChildListener mTwidereActionModeForChildListener;

    public static void setShouldNotifyChange(Activity activity) {
        if (!(activity instanceof SettingsActivity)) return;
        ((SettingsActivity) activity).setShouldNotifyChange(true);
    }

    @Override
    public void onContentChanged() {
        super.onContentChanged();
        mEntriesListView = (ListView) findViewById(R.id.entries_list);
        mSlidingPaneLayout = (SlidingPaneLayout) findViewById(R.id.sliding_pane);
        mMainContent = (TintedStatusNativeActionModeAwareLayout) findViewById(R.id.main_content);
        mDetailFragmentContainer = findViewById(R.id.detail_fragment_container);
    }


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_settings);

        final int backgroundAlpha = getCurrentThemeBackgroundAlpha();
        final int alpha = ThemeUtils.getActionBarAlpha(getThemeBackgroundOption(), backgroundAlpha);
        final int statusBarColor = ThemeUtils.getActionBarColor(this, getCurrentThemeColor(), getThemeBackgroundOption());
        mMainContent.setDrawColor(true);
        mMainContent.setColor(statusBarColor, alpha);

        mTwidereActionModeForChildListener = new TwidereActionModeForChildListener(this, this, false);
        mMainContent.setActionModeForChildListener(mTwidereActionModeForChildListener);
        mDetailFragmentContainer.setBackgroundColor((backgroundAlpha << 24 | 0xFFFFFF) & ThemeUtils.getThemeBackgroundColor(this));

        mSlidingPaneLayout.setShadowResourceLeft(R.drawable.sliding_pane_shadow_left);
        mSlidingPaneLayout.setShadowResourceRight(R.drawable.sliding_pane_shadow_right);
        mSlidingPaneLayout.setSliderFadeColor(0);
        mEntriesAdapter = new EntriesAdapter(this);
        initEntries();
        mEntriesListView.setAdapter(mEntriesAdapter);
        mEntriesListView.setChoiceMode(AbsListView.CHOICE_MODE_SINGLE);
        mEntriesListView.setOnItemClickListener(this);
    }

    private void initEntries() {
        mEntriesAdapter.addHeader(getString(R.string.appearance));
        mEntriesAdapter.addPreference(R.drawable.ic_action_color_palette, getString(R.string.theme),
                R.xml.preferences_theme);
        mEntriesAdapter.addPreference(R.drawable.ic_action_card, getString(R.string.cards),
                R.xml.preferences_cards);

        mEntriesAdapter.addHeader(getString(R.string.function));
        mEntriesAdapter.addPreference(R.drawable.ic_action_tab, getString(R.string.tabs),
                CustomTabsFragment.class);
        mEntriesAdapter.addPreference(R.drawable.ic_action_extension, getString(R.string.extensions),
                ExtensionsListFragment.class);
        mEntriesAdapter.addPreference(R.drawable.ic_action_refresh, getString(R.string.refresh),
                R.xml.preferences_refresh);
        mEntriesAdapter.addPreference(R.drawable.ic_action_notification, getString(R.string.notifications),
                R.xml.preferences_notifications);
        mEntriesAdapter.addPreference(R.drawable.ic_action_server, getString(R.string.network),
                R.xml.preferences_network);
        mEntriesAdapter.addPreference(R.drawable.ic_action_status_compose, getString(R.string.compose),
                R.xml.preferences_compose);
        mEntriesAdapter.addPreference(R.drawable.ic_action_twidere_square, getString(R.string.content),
                R.xml.preferences_content);
        mEntriesAdapter.addPreference(R.drawable.ic_action_storage, getString(R.string.storage),
                R.xml.preferences_storage);
        mEntriesAdapter.addPreference(R.drawable.ic_action_more_horizontal, getString(R.string.other_settings),
                R.xml.preferences_other);

        mEntriesAdapter.addHeader(getString(R.string.about));
        mEntriesAdapter.addPreference(R.drawable.ic_action_info, getString(R.string.about),
                R.xml.preferences_about);
        final Bundle browserArgs = new Bundle();
        browserArgs.putString(EXTRA_URI, "file:///android_asset/gpl-3.0-standalone.html");
        mEntriesAdapter.addPreference(R.drawable.ic_action_open_source, getString(R.string.open_source_license),
                SupportBrowserFragment.class, browserArgs);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (resultCode == RESULT_SETTINGS_CHANGED && data != null && data.getBooleanExtra(EXTRA_CHANGED, false)) {
            setShouldNotifyChange(true);
        }
        super.onActivityResult(requestCode, resultCode, data);
    }


    private boolean isTopSettings() {
        return Boolean.parseBoolean("true");
    }


    @Override
    public void finish() {
        if (shouldNotifyChange()) {
            final Intent data = new Intent();
            data.putExtra(EXTRA_CHANGED, true);
            setResult(isTopSettings() ? RESULT_OK : RESULT_SETTINGS_CHANGED, data);
        }
        super.finish();
    }

    private void finishNoRestart() {
        super.finish();
    }

    @Override
    public boolean handleKeyboardShortcutSingle(@NonNull KeyboardShortcutsHandler handler, int keyCode, @NonNull KeyEvent event, int metaState) {
        final String action = handler.getKeyAction(CONTEXT_TAG_NAVIGATION, keyCode, event, metaState);
        if (ACTION_NAVIGATION_BACK.equals(action)) {
            onBackPressed();
            return true;
        }
        return super.handleKeyboardShortcutSingle(handler, keyCode, event, metaState);
    }

    @Override
    public boolean isKeyboardShortcutHandled(@NonNull KeyboardShortcutsHandler handler, int keyCode, @NonNull KeyEvent event, int metaState) {
        final String action = handler.getKeyAction(CONTEXT_TAG_NAVIGATION, keyCode, event, metaState);
        return ACTION_NAVIGATION_BACK.equals(action);
    }

    @Override
    public boolean handleKeyboardShortcutRepeat(@NonNull KeyboardShortcutsHandler handler, int keyCode, int repeatCount, @NonNull KeyEvent event, int metaState) {
        return super.handleKeyboardShortcutRepeat(handler, keyCode, repeatCount, event, metaState);
    }


    private void setShouldNotifyChange(boolean notify) {
        mShouldNotifyChange = notify;
    }

    private boolean shouldNotifyChange() {
        return mShouldNotifyChange;
    }


    @Override
    public void onBackPressed() {
        if (mTwidereActionModeForChildListener.finishExisting()) {
            return;
        }
        if (isTopSettings() && shouldNotifyChange()) {
            final RestartConfirmDialogFragment df = new RestartConfirmDialogFragment();
            df.show(getSupportFragmentManager(), "restart_confirm");
            return;
        }
        super.onBackPressed();
    }

    @Override
    public boolean onPreferenceStartFragment(PreferenceFragmentCompat fragment, Preference preference) {
        final FragmentManager fm = getSupportFragmentManager();
        final FragmentTransaction ft = fm.beginTransaction();
        final Fragment f = Fragment.instantiate(this, preference.getFragment(), preference.getExtras());
        ft.replace(R.id.detail_fragment_container, f);
        ft.addToBackStack(String.valueOf(preference.getTitle()));
        ft.commit();
        return true;
    }

    @Override
    public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
        final Entry entry = mEntriesAdapter.getItem(position);
        if (!(entry instanceof PreferenceEntry)) return;
        final PreferenceEntry pe = (PreferenceEntry) entry;
        final FragmentManager fm = getSupportFragmentManager();
        final FragmentTransaction ft = fm.beginTransaction();
        if (pe.preference != 0) {
            final Bundle args = new Bundle();
            args.putInt(EXTRA_RESID, pe.preference);
            final Fragment f = Fragment.instantiate(this, SettingsDetailsFragment.class.getName(),
                    args);
            ft.replace(R.id.detail_fragment_container, f);
        } else if (pe.fragment != null) {
            ft.replace(R.id.detail_fragment_container, Fragment.instantiate(this, pe.fragment,
                    pe.args));
        }
        ft.setBreadCrumbTitle(pe.title);
        ft.commit();
        mSlidingPaneLayout.closePane();
    }

    @Override
    public boolean onPreferenceDisplayDialog(PreferenceFragmentCompat fragment, Preference preference) {
        if (preference instanceof IDialogPreference) {
            ((IDialogPreference) preference).displayDialog(fragment);
            return true;
        }
        return false;
    }

    static class EntriesAdapter extends BaseAdapter {

        static final int VIEW_TYPE_PREFERENCE_ENTRY = 0;
        static final int VIEW_TYPE_HEADER_ENTRY = 1;

        private final Context mContext;
        private final LayoutInflater mInflater;
        private final List<Entry> mEntries;

        public EntriesAdapter(Context context) {
            mContext = context;
            mInflater = LayoutInflater.from(context);
            mEntries = new ArrayList<>();
        }

        public void addPreference(@DrawableRes int icon, String title, @XmlRes int preference) {
            mEntries.add(new PreferenceEntry(icon, title, preference, null, null));
            notifyDataSetChanged();
        }

        public void addPreference(@DrawableRes int icon, String title, Class<? extends Fragment> cls) {
            addPreference(icon, title, cls, null);
        }


        public void addPreference(@DrawableRes int icon, String title, Class<? extends Fragment> cls,
                                  @Nullable Bundle args) {
            mEntries.add(new PreferenceEntry(icon, title, 0, cls.getName(), args));
            notifyDataSetChanged();
        }

        public void addHeader(String title) {
            mEntries.add(new HeaderEntry(title));
            notifyDataSetChanged();
        }

        @Override
        public int getCount() {
            return mEntries.size();
        }

        @Override
        public Entry getItem(int position) {
            return mEntries.get(position);
        }

        @Override
        public long getItemId(int position) {
            return getItem(position).hashCode();
        }

        @Override
        public boolean isEnabled(int position) {
            return getItemViewType(position) == VIEW_TYPE_PREFERENCE_ENTRY;
        }

        @Override
        public int getViewTypeCount() {
            return 2;
        }

        @Override
        public int getItemViewType(int position) {
            final Entry entry = getItem(position);
            if (entry instanceof PreferenceEntry) {
                return VIEW_TYPE_PREFERENCE_ENTRY;
            } else if (entry instanceof HeaderEntry) {
                return VIEW_TYPE_HEADER_ENTRY;
            }
            throw new UnsupportedOperationException();
        }

        @Override
        public View getView(int position, View convertView, ViewGroup parent) {
            final int viewType = getItemViewType(position);
            final Entry entry = getItem(position);
            final View view;
            if (convertView != null) {
                view = convertView;
            } else {
                switch (viewType) {
                    case VIEW_TYPE_PREFERENCE_ENTRY: {
                        view = mInflater.inflate(R.layout.list_item_preference_header_item, parent, false);
                        break;
                    }
                    case VIEW_TYPE_HEADER_ENTRY: {
                        view = mInflater.inflate(R.layout.list_item_preference_header_category, parent, false);
                        break;
                    }
                    default: {
                        throw new UnsupportedOperationException();
                    }
                }
            }
            entry.bind(view);
            return view;
        }
    }

    static abstract class Entry {
        public abstract void bind(View view);

    }

    static class PreferenceEntry extends Entry {
        private final int icon;
        private final String title;
        private final int preference;
        private final String fragment;

        private final Bundle args;

        public PreferenceEntry(int icon, String title, int preference, String fragment, Bundle args) {
            this.icon = icon;
            this.title = title;
            this.preference = preference;
            this.fragment = fragment;
            this.args = args;
        }

        @Override
        public void bind(View view) {
            ((ImageView) view.findViewById(android.R.id.icon)).setImageResource(icon);
            ((TextView) view.findViewById(android.R.id.title)).setText(title);
        }

    }

    static class HeaderEntry extends Entry {

        private final String title;

        public HeaderEntry(String title) {
            this.title = title;
        }

        @Override
        public void bind(View view) {
            ((TextView) view.findViewById(android.R.id.title)).setText(title);
        }
    }

    public static class RestartConfirmDialogFragment extends BaseSupportDialogFragment implements DialogInterface.OnClickListener {
        @NonNull
        @Override
        public Dialog onCreateDialog(Bundle savedInstanceState) {
            final AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
            builder.setMessage(R.string.app_restart_confirm);
            builder.setPositiveButton(android.R.string.ok, this);
            builder.setNegativeButton(R.string.dont_restart, this);
            return builder.create();
        }

        @Override
        public void onClick(DialogInterface dialog, int which) {
            final SettingsActivity activity = (SettingsActivity) getActivity();
            if (activity == null) return;
            switch (which) {
                case DialogInterface.BUTTON_POSITIVE: {
                    activity.finish();
                    break;
                }
                case DialogInterface.BUTTON_NEGATIVE: {
                    activity.finishNoRestart();
                    break;
                }
            }
        }
    }


}

File: twidere/src/main/kotlin/org/mariotaku/twidere/activity/SettingsActivity.kt
/*
 * 				Twidere - Twitter client for Android
 * 
 *  Copyright (C) 2012-2014 Mariotaku Lee <mariotaku.lee@gmail.com>
 * 
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 * 
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 * 
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.mariotaku.twidere.activity

import android.app.Activity
import android.app.Dialog
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.os.Bundle
import androidx.annotation.DrawableRes
import androidx.annotation.XmlRes
import androidx.fragment.app.Fragment
import androidx.core.view.ViewCompat
import androidx.appcompat.app.AlertDialog
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import androidx.preference.PreferenceFragmentCompat.OnPreferenceStartFragmentCallback
import android.view.KeyEvent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import android.widget.AdapterView.OnItemClickListener
import kotlinx.android.synthetic.main.activity_settings.*
import org.mariotaku.chameleon.Chameleon
import org.mariotaku.ktextension.Bundle
import org.mariotaku.ktextension.set
import org.mariotaku.twidere.R
import org.mariotaku.twidere.constant.IntentConstants.*
import org.mariotaku.twidere.constant.KeyboardShortcutConstants.ACTION_NAVIGATION_BACK
import org.mariotaku.twidere.constant.KeyboardShortcutConstants.CONTEXT_TAG_NAVIGATION
import org.mariotaku.twidere.extension.applyTheme
import org.mariotaku.twidere.extension.onShow
import org.mariotaku.twidere.fragment.*
import org.mariotaku.twidere.util.DeviceUtils
import org.mariotaku.twidere.util.KeyboardShortcutsHandler
import org.mariotaku.twidere.util.ThemeUtils
import java.util.*
import kotlin.system.exitProcess

class SettingsActivity : BaseActivity(), OnItemClickListener, OnPreferenceStartFragmentCallback {

    var shouldRecreate: Boolean = false
    var shouldRestart: Boolean = false
    var shouldTerminate: Boolean = false
    private lateinit var entriesAdapter: EntriesAdapter


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_settings)
        entriesAdapter = EntriesAdapter(this)

        if (savedInstanceState != null) {
            shouldRecreate = savedInstanceState.getBoolean(EXTRA_SHOULD_RECREATE, shouldRecreate)
            shouldRestart = savedInstanceState.getBoolean(EXTRA_SHOULD_RESTART, shouldRestart)
            shouldTerminate = savedInstanceState.getBoolean(EXTRA_SHOULD_TERMINATE, shouldTerminate)
        } else if (intent.getBooleanExtra(EXTRA_SHOULD_TERMINATE, false)) {
            finishNoRestart()
            exitProcess(0)
        }

        val backgroundOption = currentThemeBackgroundOption
        val backgroundAlpha = currentThemeBackgroundAlpha

        detailFragmentContainer.setBackgroundColor(ThemeUtils.getColorBackground(this,
                backgroundOption, backgroundAlpha))

        slidingPane.setShadowResourceLeft(R.drawable.sliding_pane_shadow_left)
        slidingPane.setShadowResourceRight(R.drawable.sliding_pane_shadow_right)
        slidingPane.sliderFadeColor = 0

        ViewCompat.setOnApplyWindowInsetsListener(slidingPane) listener@ { view, insets ->
            onApplyWindowInsets(view, insets)
            entriesList.setPadding(0, insets.systemWindowInsetTop, 0, insets.systemWindowInsetBottom)
            return@listener insets
        }

        initEntries()
        entriesList.adapter = entriesAdapter
        entriesList.choiceMode = AbsListView.CHOICE_MODE_SINGLE
        entriesList.onItemClickListener = this

        if (savedInstanceState == null) {
            val initialTag = intent.data?.authority
            var initialItem = -1
            var firstEntry = -1
            for (i in 0 until entriesAdapter.count) {
                val entry = entriesAdapter.getItem(i)
                if (entry is PreferenceEntry) {
                    if (firstEntry == -1) {
                        firstEntry = i
                    }
                    if (initialTag == entry.tag) {
                        initialItem = i
                        break
                    }
                }
            }
            if (initialItem == -1) {
                initialItem = firstEntry
            }
            if (initialItem != -1) {
                openDetails(initialItem)
                entriesList.setItemChecked(initialItem, true)
            }
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (resultCode == RESULT_SETTINGS_CHANGED && data != null) {
            shouldRecreate = data.getBooleanExtra(EXTRA_SHOULD_RECREATE, false)
            shouldRestart = data.getBooleanExtra(EXTRA_SHOULD_RESTART, false)
            shouldTerminate = data.getBooleanExtra(EXTRA_SHOULD_TERMINATE, false)
        }
        super.onActivityResult(requestCode, resultCode, data)
    }

    override fun finish() {
        if (shouldRecreate || shouldRestart) {
            val data = Intent()
            data.putExtra(EXTRA_SHOULD_RECREATE, shouldRecreate)
            data.putExtra(EXTRA_SHOULD_RESTART, shouldRestart)
            setResult(RESULT_SETTINGS_CHANGED, data)
        }
        super.finish()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putBoolean(EXTRA_SHOULD_RECREATE, shouldRecreate)
        outState.putBoolean(EXTRA_SHOULD_RESTART, shouldRestart)
    }

    override fun handleKeyboardShortcutSingle(handler: KeyboardShortcutsHandler, keyCode: Int, event: KeyEvent, metaState: Int): Boolean {
        val action = handler.getKeyAction(CONTEXT_TAG_NAVIGATION, keyCode, event, metaState)
        if (ACTION_NAVIGATION_BACK == action) {
            onBackPressed()
            return true
        }
        return super.handleKeyboardShortcutSingle(handler, keyCode, event, metaState)
    }

    override fun isKeyboardShortcutHandled(handler: KeyboardShortcutsHandler, keyCode: Int, event: KeyEvent, metaState: Int): Boolean {
        val action = handler.getKeyAction(CONTEXT_TAG_NAVIGATION, keyCode, event, metaState)
        return ACTION_NAVIGATION_BACK == action
    }

    override fun onSupportNavigateUp(): Boolean {
        if (notifyUnsavedChange()) {
            return true
        }
        return super.onSupportNavigateUp()
    }

    override fun onBackPressed() {
        if (notifyUnsavedChange()) return
        super.onBackPressed()
    }

    override fun onPreferenceStartFragment(fragment: PreferenceFragmentCompat, preference: Preference): Boolean {
        val fm = supportFragmentManager
        val ft = fm.beginTransaction()
        val f = Fragment.instantiate(this, preference.fragment, preference.extras)
        ft.replace(R.id.detailFragmentContainer, f)
        ft.addToBackStack(preference.title.toString())
        ft.commit()
        return true
    }

    override fun onItemClick(parent: AdapterView<*>, view: View, position: Int, id: Long) {
        openDetails(position)
    }

    private fun finishNoRestart() {
        super.finish()
    }


    private val isTopSettings: Boolean = true

    private fun initEntries() {
        entriesAdapter.addHeader(getString(R.string.appearance))
        entriesAdapter.addPreference("theme", R.drawable.ic_action_color_palette, getString(R.string.theme),
                R.xml.preferences_theme)
        entriesAdapter.addPreference("cards", R.drawable.ic_action_card, getString(R.string.cards),
                R.xml.preferences_cards)
        if (DeviceUtils.isDeviceTablet(this)) {
            entriesAdapter.addPreference("tablet_mode", R.drawable.ic_action_tablet, getString(R.string.preference_title_tablet_mode),
                    R.xml.preferences_tablet_mode)
        }

        entriesAdapter.addHeader(getString(R.string.function))
        entriesAdapter.addPreference("tabs", R.drawable.ic_action_tab, getString(R.string.tabs),
                CustomTabsFragment::class.java)
        entriesAdapter.addPreference("extension", R.drawable.ic_action_extension, getString(R.string.extensions),
                ExtensionsListFragment::class.java)
        entriesAdapter.addPreference("refresh", R.drawable.ic_action_refresh, getString(R.string.action_refresh),
                R.xml.preferences_refresh)
        entriesAdapter.addPreference("notifications", R.drawable.ic_action_notification, getString(R.string.settings_notifications),
                R.xml.preferences_notifications)
        entriesAdapter.addPreference("network", R.drawable.ic_action_web, getString(R.string.network),
                R.xml.preferences_network)
        entriesAdapter.addPreference("compose", R.drawable.ic_action_status_compose, getString(R.string.action_compose),
                R.xml.preferences_compose)
        entriesAdapter.addPreference("content", R.drawable.ic_action_twidere_square, getString(R.string.content),
                R.xml.preferences_content)
        entriesAdapter.addPreference("storage", R.drawable.ic_action_storage, getString(R.string.preference_title_storage),
                R.xml.preferences_storage)
        entriesAdapter.addPreference("other", R.drawable.ic_action_more_horizontal, getString(R.string.other_settings),
                R.xml.preferences_other)

        entriesAdapter.addHeader(getString(R.string.title_about))
        entriesAdapter.addPreference("about", R.drawable.ic_action_info, getString(R.string.title_about),
                R.xml.preferences_about)
        val browserArgs = Bundle()
        browserArgs.putString(EXTRA_URI, "file:///android_asset/gpl-3.0-standalone.html")
        entriesAdapter.addPreference("license", R.drawable.ic_action_open_source, getString(R.string.title_open_source_license),
                BrowserFragment::class.java, browserArgs)
    }

    private fun openDetails(position: Int) {
        if (isFinishing) return
        val entry = entriesAdapter.getItem(position) as? PreferenceEntry ?: return
        val fm = supportFragmentManager
        fm.popBackStackImmediate(null, 0)
        val ft = fm.beginTransaction()
        if (entry.preference != 0) {
            val args = Bundle()
            args.putInt(EXTRA_RESID, entry.preference)
            val f = Fragment.instantiate(this, SettingsDetailsFragment::class.java.name,
                    args)
            ft.replace(R.id.detailFragmentContainer, f)
        } else if (entry.fragment != null) {
            ft.replace(R.id.detailFragmentContainer, Fragment.instantiate(this, entry.fragment,
                    entry.args))
        }
        ft.setBreadCrumbTitle(entry.title)
        ft.commit()
        slidingPane.closePane()
    }

    private fun notifyUnsavedChange(): Boolean {
        if (isTopSettings && (shouldRecreate || shouldRestart || shouldTerminate)) {
            executeAfterFragmentResumed {
                if (it.isFinishing) return@executeAfterFragmentResumed
                it as SettingsActivity
                val df = RestartConfirmDialogFragment()
                df.arguments = Bundle {
                    this[EXTRA_SHOULD_RECREATE] = it.shouldRecreate
                    this[EXTRA_SHOULD_RESTART] = it.shouldRestart
                    this[EXTRA_SHOULD_TERMINATE] = it.shouldTerminate
                }
                df.show(it.supportFragmentManager, "restart_confirm")
            }
            return true
        }
        return false
    }

    internal class EntriesAdapter(context: Context) : BaseAdapter() {
        private val inflater: LayoutInflater = LayoutInflater.from(context)
        private val entries: MutableList<Entry> = ArrayList()

        fun addPreference(tag: String, @DrawableRes icon: Int, title: String, @XmlRes preference: Int) {
            entries.add(PreferenceEntry(tag, icon, title, preference, null, null))
            notifyDataSetChanged()
        }


        fun addPreference(tag: String, @DrawableRes icon: Int, title: String, cls: Class<out Fragment>,
                          args: Bundle? = null) {
            entries.add(PreferenceEntry(tag, icon, title, 0, cls.name, args))
            notifyDataSetChanged()
        }

        fun addHeader(title: String) {
            entries.add(HeaderEntry(title))
            notifyDataSetChanged()
        }

        override fun getCount(): Int {
            return entries.size
        }

        override fun getItem(position: Int): Entry {
            return entries[position]
        }

        override fun getItemId(position: Int): Long {
            return getItem(position).hashCode().toLong()
        }

        override fun isEnabled(position: Int): Boolean {
            return getItemViewType(position) == VIEW_TYPE_PREFERENCE_ENTRY
        }

        override fun getViewTypeCount(): Int {
            return 2
        }

        override fun getItemViewType(position: Int): Int {
            val entry = getItem(position)
            if (entry is PreferenceEntry) {
                return VIEW_TYPE_PREFERENCE_ENTRY
            } else if (entry is HeaderEntry) {
                return VIEW_TYPE_HEADER_ENTRY
            }
            throw UnsupportedOperationException()
        }

        override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
            val viewType = getItemViewType(position)
            val entry = getItem(position)
            val view: View = convertView ?: let {
                when (viewType) {
                    VIEW_TYPE_PREFERENCE_ENTRY -> {
                        return@let inflater.inflate(R.layout.list_item_preference_header_item, parent, false)
                    }
                    VIEW_TYPE_HEADER_ENTRY -> {
                        return@let inflater.inflate(R.layout.list_item_preference_header_category, parent, false)
                    }
                    else -> {
                        throw UnsupportedOperationException()
                    }
                }
            }
            entry.bind(view)
            return view
        }

        companion object {

            const val VIEW_TYPE_PREFERENCE_ENTRY = 0
            const val VIEW_TYPE_HEADER_ENTRY = 1
        }
    }

    internal abstract class Entry {
        abstract fun bind(view: View)

    }

    internal class PreferenceEntry(
            val tag: String,
            val icon: Int,
            val title: String,
            val preference: Int,
            val fragment: String?,
            val args: Bundle?
    ) : Entry() {

        override fun bind(view: View) {
            view.findViewById<ImageView>(android.R.id.icon).setImageResource(icon)
            view.findViewById<TextView>(android.R.id.title).text = title
        }

    }

    internal class HeaderEntry(private val title: String) : Entry() {

        override fun bind(view: View) {
            val theme = Chameleon.getOverrideTheme(view.context, view.context)
            val textView: TextView = view.findViewById(android.R.id.title)
            textView.setTextColor(ThemeUtils.getOptimalAccentColor(theme.colorAccent,
                    theme.colorForeground))
            textView.text = title
        }
    }

    class RestartConfirmDialogFragment : BaseDialogFragment(), DialogInterface.OnClickListener {
        override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
            val builder = AlertDialog.Builder(requireActivity())
            if (arguments?.getBoolean(EXTRA_SHOULD_TERMINATE) == true) {
                builder.setMessage(R.string.app_terminate_confirm)
                builder.setNegativeButton(R.string.action_dont_terminate, this)
            } else {
                builder.setMessage(R.string.app_restart_confirm)
                builder.setNegativeButton(R.string.action_dont_restart, this)
            }
            builder.setPositiveButton(android.R.string.ok, this)
            val dialog = builder.create()
            dialog.onShow { it.applyTheme() }
            return dialog
        }

        override fun onClick(dialog: DialogInterface, which: Int) {
            val activity = activity as SettingsActivity
            when (which) {
                DialogInterface.BUTTON_POSITIVE -> {
                    if (arguments?.getBoolean(EXTRA_SHOULD_TERMINATE) == true) {
                        val intent = Intent(context, SettingsActivity::class.java)
                        intent.putExtra(EXTRA_SHOULD_TERMINATE, true)
                        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK or Intent.FLAG_ACTIVITY_NEW_TASK)
                        activity.startActivity(intent)
                    } else {
                        activity.finish()
                    }
                }
                DialogInterface.BUTTON_NEGATIVE -> {
                    activity.finishNoRestart()
                }
            }
        }
    }

    companion object {

        private const val RESULT_SETTINGS_CHANGED = 10

        fun setShouldRecreate(activity: Activity) {
            if (activity !is SettingsActivity) return
            activity.shouldRecreate = true
        }

        fun setShouldRestart(activity: Activity) {
            if (activity !is SettingsActivity) return
            activity.shouldRestart = true
        }

        fun setShouldTerminate(activity: Activity) {
            if (activity !is SettingsActivity) return
            activity.shouldTerminate = true
        }
    }


}
